### ISSUE TITLE
It looks roughly to set the search completely fail when we use sql search and partial shards search fail

### ISSUE THREAD

**Issue by kkewwei** [elastic/elasticsearch#82531]
We know that the response of dsl search will show the right result and failures reason if partial shards fail, as follows:
```
{
   "took": 117,
   "timed_out": false,
   "_shards": {
      "total": 10,
      "successful": 5,
      "skipped": 0,
      "failed": 5,
      "failures": [
         {
            "shard": 0,
            "index": "test5",
            "node": "Y74ImDn6RGyxv_e9e3uR9Q",
            "reason": {
               "type": "query_shard_exception",
               "reason": "failed to create query: Cannot search on field [field1] since it is not indexed.",
               "index_uuid": "F1Jfl1IeQli23M198UyRuA",
               "index": "test5",
               "caused_by": {
                  "type": "illegal_argument_exception",
                  "reason": "Cannot search on field [field1] since it is not indexed."
               }
            }
         }
      ]
   },
   "hits": {
      "total": {
         "value": 1,
         "relation": "eq"
      },
      "max_score": 0.2876821,
      "hits": [
         {
            "_index": "test6",
            "_type": "_doc",
            "_id": "1",
            "_score": 0.2876821,
            "_source": {
               "keyword_false": "1"
            }
         }
      ]
   }
}
```
When I use sql api to search, even if one shard fails to search, es will think the search as completely failure, discard partial correct results. the fail response is as follows:
```
{
  "error": {
    "root_cause": [
      {
        "type": "query_shard_exception",
        "reason": "failed to create query: Cannot search on field [field1] since it is not indexed.",
        "index_uuid": "F1Jfl1IeQli23M198UyRuA",
        "index": "test5"
      }
    ],
    "type": "sql_illegal_argument_exception",
    "reason": "RemoteTransportException[[node-1][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: QueryShardException[failed to create query: Cannot search on field [field1] since it is not indexed.]; nested: IllegalArgumentException[Cannot search on field [field1] since it is not indexed.]; ",
    "caused_by": {
      "type": "query_shard_exception",
      "reason": "failed to create query: Cannot search on field [field1] since it is not indexed.",
      "index_uuid": "F1Jfl1IeQli23M198UyRuA",
      "index": "test5",
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Cannot search on field [field1] since it is not indexed."
      }
    }
  },
  "status": 500
}
```
(see https://github.com/elastic/elasticsearch/blob/ebcea81687fbe280e330f611317e051c5cb0cdf0/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java#L602)

It't common in product that some of shards fail to search, we can tolerate partial failure, sometimes partial shard search failure is expected.

If we could match it with dsl syntax, or we can add parameter to deceide  only return the partial correct result.

**Comment by kkewwei** [comment]
@not-napoleon can you please comfirm it?

**Comment by bpintea** [comment]
Could you please elaborate on what practical gain would your proposition bring, @kkewwei?
What is the context in which this would be useful? (querying index patterns covering indices with different mappings?)
Such an extension might work with search hits (simple data loading), but aggregations would fail or worse, return erroneous data.
Why not correct the query to exclude the offending field?

**Comment by kkewwei** [comment]
We use in log scenarios, create new index every day, and query all daily indexes through aliases. At first we think the field is useless and set it `not indexed`, later, the field become important and set it `indexed`, but we can't query the field by sql unless the `not indexed` logs is expire.

It's common case, It's expected that searching `not indexed` log will fail, as long as the new field can be queried. Aggregates are used less often than queries. It's  acceptable that aggregates are not available.


It's not worth the loss to exclude the offending field, in that way, we will no longer be able to use alias, and has to deal with specific indices.

**Comment by bpintea** [comment]
Related: #32284, #33148, #77547.

**Comment by bpintea** [comment]
Fixed by #85897.
